src/team_lib/mhba/xccl_mhba_mkeys.c

```diff
@@ -0,0 +1,262 @@
+        return XCCL_ERR_NO_MESSAGE;
+    }
+    xccl_mhba_debug("umr_master_key_dv_mkey: lkey=0x%x", mkey->lkey);
+    *mkey_ptr = mkey;
+    return XCCL_OK;
+}
+
+xccl_status_t xccl_mhba_init_mkeys(xccl_mhba_context_t *ctx, xccl_mhba_node_t *node) {
+    xccl_status_t status;
+    status = create_matser_key(ctx, node, &node->send_mkey);
+    if (status != XCCL_OK) {
+        xccl_mhba_error("create send masterkey failed");
+        return status;
+    }
+    status = create_matser_key(ctx, node, &node->recv_mkey);
+    if (status != XCCL_OK) {
+        xccl_mhba_error("create recv masterkey failed");
+        if (mlx5dv_destroy_mkey(node->send_mkey)){
+            xccl_mhba_error("mkey destroy failed (errno=%d)", errno);
+        }
+        return status;
+    }
+    return XCCL_OK;
+}
+
+static xccl_status_t poll_umr_cq(xccl_mhba_context_t *ctx) {
+    int expected_completions_num = 1;
+    struct ibv_wc work_completion[1];
+    memset(work_completion, 0, expected_completions_num * sizeof(struct ibv_wc));
+    int num_completions;
+    xccl_status_t status = poll_cq(ctx->umr_cq, expected_completions_num, work_completion);
+    if (status != XCCL_OK) {
+        xccl_mhba_error("UMR WQE on UMR QP failed");
+        return status;
+    }
+    if (work_completion[0].status != IBV_WC_SUCCESS) {
+        xccl_mhba_error("UMR CQ returned completion with status %s (%d)",
+                        ibv_wc_status_str(work_completion[0].status), work_completion[0].status);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+    if (work_completion[0].opcode != MLX5DV_WC_UMR) {
+        xccl_mhba_error("Got unexpected completion opcode for the UMR QP: %d", work_completion[0].opcode);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+    xccl_mhba_debug("Successfully executed the UMR WQE");
+    return XCCL_OK;
+}
+
+// Execute the UMR WQE for populating the UMR's MasterMKey
+static xccl_status_t
+populate_mkey(xccl_mhba_context_t *ctx, xccl_mhba_node_t *node, int mem_access_flags, struct mlx5dv_mkey *mkey,
+              void *mkey_entries) {
+    xccl_status_t status;
+    xccl_mhba_debug("Execute the UMR WQE for populating the MasterMKeys");
+    ibv_wr_start(ctx->umr_qpx);
+    ctx->umr_qpx->wr_id = 1; // First (and only) WR todo: check meaning
+    int repeat_count = node->sbgp->group_size / node->block_size; // todo check calc
+    mlx5dv_wr_mr_interleaved(ctx->umr_mlx5dv_qp_ex, mkey, mem_access_flags,
+                             repeat_count, node->sbgp->group_size,
+                             (struct mlx5dv_mr_interleaved *) mkey_entries); // todo check for error
+    status = poll_umr_cq(ctx);
+    if (status != XCCL_OK) {
+        return status;
+    }
+    return XCCL_OK;
+}
+
+xccl_status_t xccl_mhba_populate_send_recv_mkeys(xccl_mhba_context_t *ctx, xccl_mhba_node_t *node){
+    xccl_status_t status;
+    int send_mem_access_flags = 0;
+    status = populate_mkey(ctx,node,send_mem_access_flags,node->send_mkey,node->my_send_umr_data);
```

Comment:
why "my_send_umr_data" is used? It shuold be just "send_umr_data" that points to the beginning of the mr_interleaved array (if ASR rank is not 0 it will not be the same as "my_umr_data")

Reply:
good catch